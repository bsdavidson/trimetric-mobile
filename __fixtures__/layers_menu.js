export const display = true;

export const bottom = 60;

export const selectedItemsViewHeight = 0;

export const layerVisibility = {
  routeShapes: true,
  buses: false,
  trains: true,
  stops: true,
  vehicleLabels: true,
  buildings: false
};

export const dimensions = {
  window: {fontScale: 1, width: 375, height: 667, scale: 2},
  screen: {fontScale: 1, width: 375, height: 667, scale: 2}
};
