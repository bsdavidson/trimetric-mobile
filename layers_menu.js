import React, {Component} from "react";
import {
  Image,
  PixelRatio,
  StyleSheet,
  Switch,
  Text,
  TouchableOpacity,
  View
} from "react-native";
import {connect} from "react-redux";

import {updateLayerVisibility} from "./actions";
import {HEADER_CLOSED_MAX_HEIGHT} from "./selected_items_view";
import {STAT_MENU_HEIGHT} from "./stat_menu";

import layersIcon from "./assets/layers.png";

const CONTAINER_PADDING = 15;
const MIN_SCREEN_HEIGHT = 400;
const OPTION_MARGIN = 20;
const OPTION_WIDTH = 170;
const OPTIONS = [
  {label: "Route Lines", key: "routeShapes"},
  {label: "Stops", key: "stops"},
  {label: "Buses", key: "buses"},
  {label: "Trains", key: "trains"},
  {label: "Vehicle Labels", key: "vehicleLabels"},
  {label: "3D Buildings", key: "buildings"}
];

export class LayersMenu extends Component {
  state = {
    visible: this.props.openState || false
  };

  constructor(props) {
    super(props);

    this.handleTogglePress = this.handleTogglePress.bind(this);
  }

  handleTogglePress() {
    this.setState({visible: !this.state.visible});
  }

  render() {
    // Viewheight > HEADER_CLOSED_MAX_HEIGHT means it is in an open state, so do not render the
    // layers menu as there probably won't be enough room
    if (this.props.selectedItemsViewHeight > HEADER_CLOSED_MAX_HEIGHT) {
      return null;
    }
    let bottom = Math.max(
      this.props.selectedItemsViewHeight,
      PixelRatio.getPixelSizeForLayoutSize(STAT_MENU_HEIGHT)
    );
    if (!this.state.visible) {
      return (
        <TouchableOpacity
          onPress={this.handleTogglePress}
          style={[styles.button, {bottom}]}>
          <Image style={styles.buttonIcon} source={layersIcon} />
        </TouchableOpacity>
      );
    }
    const {height} = this.props.dimensions.screen;
    const columns = height < MIN_SCREEN_HEIGHT ? 2 : 1;
    const width = CONTAINER_PADDING + (OPTION_WIDTH + OPTION_MARGIN) * columns;

    return (
      <View style={[styles.container, {width, bottom: bottom + 20}]}>
        <TouchableOpacity
          style={styles.closeButton}
          onPress={this.handleTogglePress}>
          <Text style={styles.closeButtonIcon}>&#x00D7;</Text>
        </TouchableOpacity>

        <Text style={styles.title}>Visible Features</Text>
        <View style={styles.optionContainer}>
          {OPTIONS.map((o, i) => (
            <View key={i} style={styles.option}>
              <Text style={styles.optionTitle}>{o.label}</Text>
              <Switch
                value={this.props.layerVisibility[o.key]}
                style={styles.optionSwitch}
                onValueChange={e => {
                  this.props.onUpdateLayerVisibility(o.key, e);
                }}
              />
            </View>
          ))}
        </View>
      </View>
    );
  }
}

function mapStateToProps(state) {
  return {
    dimensions: state.dimensions,
    layerVisibility: state.layerVisibility,
    selectedItemsViewHeight: state.selectedItemsViewHeight
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onUpdateLayerVisibility: (layerName, value) => {
      dispatch(updateLayerVisibility(layerName, value));
    }
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(LayersMenu);

const styles = StyleSheet.create({
  container: {
    backgroundColor: "#FFFFFF",
    borderRadius: 6,
    elevation: 4,
    maxHeight: "100%",
    padding: CONTAINER_PADDING,
    paddingRight: 0,
    position: "absolute",
    right: 35,
    shadowColor: "#000",
    shadowOffset: {width: 0, height: 4},
    shadowOpacity: 0.4,
    shadowRadius: 6
  },
  button: {
    margin: 5,
    marginRight: 0,
    padding: 5,
    paddingRight: 0,
    position: "absolute",
    right: -8,
    zIndex: 90
  },
  buttonIcon: {
    height: 66,
    opacity: 0.6,
    shadowColor: "#000",
    shadowOffset: {width: 0, height: 4},
    shadowOpacity: 0.2,
    shadowRadius: 4,
    width: 66
  },
  title: {
    color: "#AAAAAA",
    fontSize: 14
  },
  option: {
    alignItems: "center",
    flexDirection: "row",
    marginRight: OPTION_MARGIN,
    marginTop: 10,
    width: OPTION_WIDTH
  },
  optionContainer: {
    flexDirection: "row",
    flexWrap: "wrap"
  },
  optionTitle: {
    color: "#666667",
    flex: 1
  },
  optionSwitch: {},
  closeButton: {
    flex: 0,
    position: "absolute",
    right: 0,
    zIndex: 99
  },
  closeButtonIcon: {
    color: "#AAAAAA",
    fontSize: 30,
    fontWeight: "bold",
    lineHeight: 30,
    margin: 5,
    marginRight: 8,
    padding: 2,
    paddingLeft: 6,
    paddingRight: 6
  }
});
