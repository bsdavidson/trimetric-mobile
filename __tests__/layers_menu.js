import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import {LayersMenu} from "../layers_menu";
import {
  bottom,
  dimensions,
  display,
  layerVisibility,
  selectedItemsViewHeight
} from "../__fixtures__/layers_menu";

it("renders layers menu", () => {
  // If the bottom menu view Height is over 150px, do
  // not render the layers menu.
  const nullLayersMenu = renderer
    .create(
      <LayersMenu
        bottom={bottom}
        dimensions={dimensions}
        display={display}
        layerVisibility={layerVisibility}
        openState={false}
        selectedItemsViewHeight={200}
      />
    )
    .toJSON();
  expect(layersMenuClosed).toMatchSnapshot("null_layers_menu");

  const layersMenuClosed = renderer
    .create(
      <LayersMenu
        bottom={bottom}
        dimensions={dimensions}
        display={display}
        layerVisibility={layerVisibility}
        openState={false}
        selectedItemsViewHeight={selectedItemsViewHeight}
      />
    )
    .toJSON();
  expect(layersMenuClosed).toMatchSnapshot("layers_menu_closed");

  const layersMenuOpen = renderer
    .create(
      <LayersMenu
        bottom={bottom}
        dimensions={dimensions}
        display={display}
        layerVisibility={layerVisibility}
        openState={true}
        selectedItemsViewHeight={selectedItemsViewHeight}
      />
    )
    .toJSON();
  expect(layersMenuOpen).toMatchSnapshot("layers_menu_open");
});
