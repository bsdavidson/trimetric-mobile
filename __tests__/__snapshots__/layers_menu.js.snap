// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layers_menu_closed 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "bottom": 80,
      "margin": 5,
      "marginRight": 0,
      "opacity": 1,
      "padding": 5,
      "paddingRight": 0,
      "position": "absolute",
      "right": -8,
      "zIndex": 90,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Image
    source={
      Object {
        "testUri": "../../../assets/layers.png",
      }
    }
    style={
      Object {
        "height": 66,
        "opacity": 0.6,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 4,
        "width": 66,
      }
    }
  />
</View>
`;

exports[`layers_menu_open 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 6,
        "elevation": 4,
        "maxHeight": "100%",
        "padding": 15,
        "paddingRight": 0,
        "position": "absolute",
        "right": 35,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "shadowRadius": 6,
      },
      Object {
        "bottom": 100,
        "width": 205,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "zIndex": 99,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#AAAAAA",
          "fontSize": 30,
          "fontWeight": "bold",
          "lineHeight": 30,
          "margin": 5,
          "marginRight": 8,
          "padding": 2,
          "paddingLeft": 6,
          "paddingRight": 6,
        }
      }
    >
      Ã—
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#AAAAAA",
        "fontSize": 14,
      }
    }
  >
    Visible Features
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        Route Lines
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={true}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        Stops
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={true}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        Buses
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={false}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        Trains
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={true}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        Vehicle Labels
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={true}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 20,
          "marginTop": 10,
          "width": 170,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#666667",
            "flex": 1,
          }
        }
      >
        3D Buildings
      </Text>
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {},
          ]
        }
        value={false}
      />
    </View>
  </View>
</View>
`;

exports[`null_layers_menu 1`] = `undefined`;
